File: android/app/src/main/kotlin/com/poyal/drawry/AppModel.kt
package com.poyal.drawry

import android.app.Application
import android.net.Uri
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.poyal.drawry.core.*
import com.poyal.drawry.data.*
import java.io.File
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.*

data class AppState(
    val ready: Boolean = false,
    val diaries: List<Diary> = emptyList(),
    val preferences: Preferences = Preferences(),
    val draft: Diary? = null,
    val editor: Diary? = null,
    val busy: Boolean = false,
    val canCancel: Boolean = false,
    val error: String? = null,
    val locked: Boolean = true,
    val restore: Pair<Snapshot, File>? = null,
    val erased: Boolean = false,
)

class AppModel(application: Application) : AndroidViewModel(application) {
    private val mutable = MutableStateFlow(AppState())
    val state = mutable.asStateFlow()
    lateinit var vault: Vault
        private set

    lateinit var media: MediaStore
        private set

    private var autosave: Job? = null
    private var operation: Job? = null

    init {
        action {
            vault = withContext(Dispatchers.IO) { Vault(application) }
            media = MediaStore(application, vault)
            vault.initialize()
            refresh()
            mutable.update { it.copy(ready = true, locked = it.preferences.lockEnabled) }
        }
    }

    fun error(message: String?) {
        mutable.update { it.copy(error = message) }
    }

    fun action(cancellable: Boolean = false, block: suspend () -> Unit) {
        if (mutable.value.busy) return
        operation =
            viewModelScope.launch {
                mutable.update { it.copy(busy = true, canCancel = cancellable) }
                try {
                    block()
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    error(e.message ?: "작업을 완료하지 못했습니다. 다시 시도해 주세요.")
                } finally {
                    mutable.update { it.copy(busy = false, canCancel = false) }
                }
            }
    }

    fun cancelOperation() {
        if (mutable.value.canCancel) operation?.cancel()
    }

    suspend fun refresh() {
        val diaries = vault.diaries()
        val preferences = vault.preferences()
        val draft = vault.draft()
        mutable.update { it.copy(diaries = diaries, preferences = preferences, draft = draft) }
    }

    fun preferences(value: Preferences) = action {
        vault.preferences(value)
        mutable.update { it.copy(preferences = value) }
    }

    fun startEditor(diary: Diary? = null) {
        mutable.update { it.copy(editor = diary ?: it.draft ?: Diary()) }
    }

    fun updateEditor(value: Diary) {
        mutable.update { it.copy(editor = value) }
        autosave?.cancel()
        autosave =
            viewModelScope.launch {
                delay(300)
                try {
                    vault.draft(value)
                    mutable.update { it.copy(draft = value) }
                } catch (e: CancellationException) {
                    throw e
                } catch (e: Exception) {
                    error("초안을 저장하지 못했습니다.")
                }
            }
    }

    fun leaveEditor(discard: Boolean) = action {
        autosave?.cancelAndJoin()
        if (discard) vault.draft(null) else mutable.value.editor?.let { vault.draft(it) }
        mutable.update { it.copy(editor = null) }
        refresh()
    }

    fun saveEditor() = action {
        autosave?.cancelAndJoin()
        vault.save(requireNotNull(mutable.value.editor))
        mutable.update { it.copy(editor = null) }
        refresh()
    }

    fun import(uris: List<Uri>, preserve: Boolean) =
        action(cancellable = true) {
            for (uri in uris) {
                val editor = mutable.value.editor ?: break
                require(editor.media.size < 10) { MEDIA_LIMIT }
                val video =
                    getApplication<Application>()
                        .contentResolver
                        .getType(uri)
                        ?.startsWith("video/") == true
                val item = media.import(uri, video, preserve)
                val updated = editor.copy(media = editor.media + item)
                try {
                    updated.validate()
                    autosave?.cancelAndJoin()
                    vault.draft(updated)
                    mutable.update { it.copy(editor = updated, draft = updated) }
                } catch (e: Exception) {
                    item.files.forEach { File(vault.mediaDirectory, it).delete() }
                    throw e
                }
            }
        }

    fun editMedia(item: Media, recipe: EditRecipe) = action {
        replaceMedia(media.edit(item, recipe))
    }

    fun frame(item: Media, millis: Long) = action { replaceMedia(media.frame(item, millis)) }

    private suspend fun replaceMedia(item: Media) {
        val editor = mutable.value.editor ?: return
        val next = editor.copy(media = editor.media.map { if (it.id == item.id) item else it })
        autosave?.cancelAndJoin()
        vault.draft(next)
        mutable.update { it.copy(editor = next, draft = next) }
    }

    fun trash(diary: Diary, restore: Boolean = false) = action {
        vault.trash(diary, restore)
        refresh()
    }

    fun purge(diary: Diary) = action {
        vault.purge(diary.id)
        refresh()
    }

    fun inspect(uri: Uri, password: String, recovery: String) =
        action(cancellable = true) {
            val source = File(vault.cache, "import.drawry")
            try {
                withContext(Dispatchers.IO) {
                    getApplication<Application>().contentResolver.openInputStream(uri)!!.use { input
                        ->
                        source.outputStream().use { input.copyTo(it) }
                    }
                }
                val key =
                    if (recovery.isBlank()) null
                    else java.util.Base64.getDecoder().decode(recovery.trim())
                val result = vault.inspect(source, password.ifEmpty { null }, key)
                mutable.update { it.copy(restore = result) }
            } finally {
                source.delete()
            }
        }

    fun restore(replace: Boolean) = action {
        val pending = mutable.value.restore ?: return@action
        vault.restore(pending.first, pending.second, replace)
        mutable.update { it.copy(restore = null) }
        refresh()
    }

    fun cancelRestore() {
        mutable.value.restore?.second?.deleteRecursively()
        mutable.update { it.copy(restore = null) }
    }

    fun unlock() {
        mutable.update { it.copy(locked = false) }
    }

    fun lock() {
        if (mutable.value.preferences.lockEnabled) mutable.update { it.copy(locked = true) }
    }

    fun background() {
        if (::media.isInitialized) media.suspendViewing()
        viewModelScope.launch {
            mutable.value.editor?.let { runCatching { vault.draft(it) } }
            if (::media.isInitialized) withContext(Dispatchers.IO) { media.clearViewing() }
        }
    }

    fun foreground() {
        if (::media.isInitialized) media.allowViewing()
        if (::vault.isInitialized && mutable.value.ready && !mutable.value.busy)
            action {
                withContext(Dispatchers.IO) { vault.cleanExpiredExports() }
                vault.purgeExpired()
                refresh()
            }
    }

    fun erase() = action {
        withContext(Dispatchers.IO) { vault.erase() }
        mutable.value = AppState(erased = true, ready = true, locked = false)
    }
}

File: ios/App/AppModel.swift
import DrawryCore
import Foundation
import LocalAuthentication
import Observation

@MainActor @Observable final class AppModel {
  var vault: Vault?
  var media: MediaStore?
  var diaries: [Diary] = []
  var preferences = Preferences()
  var editor: Diary?
  var draft: Diary?
  var busy = false
  var canCancel = false
  var ready = false
  var locked = true
  var covered = false
  var error: String?
  var erased = false
  var pendingRestore: (Snapshot, URL)?
  var sharedFile: URL?
  private var autosave: Task<Void, Never>?
  private var operation: Task<Void, Never>?
  private var privacyTask: Task<Void, Never>?
  private var backgroundAt: TimeInterval?
  init() { run { try await self.open() } }
  func open() async throws {
    let vault = try Vault()
    self.vault = vault
    media = MediaStore(vault: vault)
    try await vault.initialize()
    try await refresh()
    locked = preferences.lockEnabled
    ready = true
  }
  func run(cancellable: Bool = false, _ work: @escaping @MainActor () async throws -> Void) {
    guard !busy else { return }
    busy = true
    canCancel = cancellable
    operation = Task {
      defer {
        busy = false
        canCancel = false
      }
      do { try await work() } catch is CancellationError {} catch {
        self.error = error.localizedDescription
      }
    }
  }
  func cancelOperation() { if canCancel { operation?.cancel() } }
  func refresh() async throws {
    guard let vault else { return }
    diaries = try await vault.diaries()
    preferences = try await vault.preferences()
    draft = try await vault.draft()
  }
  func setPreferences(_ value: Preferences) {
    run {
      try await self.vault?.preferences(value)
      self.preferences = value
    }
  }
  func startEditor(_ diary: Diary? = nil) { editor = diary ?? draft ?? Diary() }
  func update(_ value: Diary) {
    editor = value
    autosave?.cancel()
    autosave = Task {
      do {
        try await Task.sleep(for: .milliseconds(300))
        try Task.checkCancellation()
        try await vault?.draft(value)
        draft = value
      } catch is CancellationError {} catch { self.error = "초안을 저장하지 못했습니다." }
    }
  }
  func leaveEditor(discard: Bool) {
    run {
      self.autosave?.cancel()
      await self.autosave?.value
      try await self.vault?.draft(discard ? nil : self.editor)
      self.editor = nil
      try await self.refresh()
    }
  }
  func save() {
    run {
      self.autosave?.cancel()
      await self.autosave?.value
      guard let value = self.editor else { return }
      try await self.vault?.save(value)
      self.editor = nil
      try await self.refresh()
    }
  }
  func importFiles(_ files: [(URL, Bool)], preserve: Bool) {
    run(cancellable: true) {
      defer { for (url, _) in files { try? FileManager.default.removeItem(at: url) } }
      guard let media = self.media, let vault = self.vault else { return }
      for (url, video) in files {
        guard var value = self.editor else { return }
        guard value.media.count < 10 else { throw VaultError.mediaLimit }
        let item = try await media.importFile(url, video: video, preserve: preserve)
        value.media.append(item)
        do {
          try value.validate()
          self.autosave?.cancel()
          await self.autosave?.value
          try await vault.draft(value)
          self.editor = value
          self.draft = value
        } catch {
          for name in item.files {
            try? FileManager.default.removeItem(
              at: await vault.mediaDirectory.appendingPathComponent(name))
          }
          throw error
        }
      }
    }
  }
  func edit(_ item: Media, recipe: EditRecipe) {
    run {
      guard let media = self.media else { return }
      try await self.replaceMedia(media.edit(item, recipe: recipe))
    }
  }
  func frame(_ item: Media, millis: Int64) {
    run {
      guard let media = self.media else { return }
      try await self.replaceMedia(media.frame(item, millis: millis))
    }
  }
  private func replaceMedia(_ value: Media) async throws {
    guard var editor else { return }
    editor.media = editor.media.map { $0.id == value.id ? value : $0 }
    autosave?.cancel()
    await autosave?.value
    try await vault?.draft(editor)
    self.editor = editor
    self.draft = editor
  }
  func trash(_ diary: Diary, restore: Bool = false) {
    run {
      try await self.vault?.trash(diary, restore: restore)
      try await self.refresh()
    }
  }
  func purge(_ diary: Diary) {
    run {
      try await self.vault?.purge(diary.id)
      try await self.refresh()
    }
  }
  func authenticate(_ success: @escaping @MainActor () -> Void) {
    let context = LAContext()
    var failure: NSError?
    guard context.canEvaluatePolicy(.deviceOwnerAuthentication, error: &failure) else {
      error = "기기 설정에서 암호 또는 생체 인증을 먼저 설정해 주세요."
      return
    }
    Task {
      do {
        if try await context.evaluatePolicy(
          .deviceOwnerAuthentication, localizedReason: "나만의 Drawry 기록을 확인합니다.")
        {
          success()
        }
      } catch { self.error = error.localizedDescription }
    }
  }
  func background() {
    covered = true
    if backgroundAt == nil { backgroundAt = ProcessInfo.processInfo.systemUptime }
    privacyTask = Task {
      await media?.clearViewing()
      if let editor { try? await vault?.draft(editor) }
    }
  }
  func foreground() {
    Task {
      await privacyTask?.value
      await media?.allowViewing()
    }
    if preferences.lockEnabled, let time = backgroundAt,
      ProcessInfo.processInfo.systemUptime - time >= Double(preferences.graceSeconds)
    {
      locked = true
    }
    backgroundAt = nil
    covered = false
    if ready && !busy {
      run {
        try await self.vault?.purgeExpired()
        try await self.refresh()
      }
    }
  }
  func inspect(_ url: URL, password: String, recovery: String) {
    run(cancellable: true) {
      guard let vault = self.vault else { return }
      let scoped = url.startAccessingSecurityScopedResource()
      defer { if scoped { url.stopAccessingSecurityScopedResource() } }
      self.pendingRestore = try await vault.inspect(
        source: url, password: password.isEmpty ? nil : password,
        recovery: recovery.isEmpty
          ? nil : Data(base64Encoded: recovery.trimmingCharacters(in: .whitespacesAndNewlines)))
    }
  }
  func restore(replace: Bool) {
    guard let pending = pendingRestore, !busy else { return }
    pendingRestore = nil
    run {
      try await self.vault?.restore(snapshot: pending.0, stage: pending.1, replace: replace)
      try await self.refresh()
    }
  }
  func cancelRestore() {
    if let pendingRestore { try? FileManager.default.removeItem(at: pendingRestore.1) }
    pendingRestore = nil
  }
  func erase() {
    authenticate {
      self.run {
        try await self.vault?.erase()
        self.diaries = []
        self.editor = nil
        self.draft = nil
        self.erased = true
        self.locked = false
      }
    }
  }
}
